pricegraph/src/lib.rs

```diff
@@ -41,3 +44,123 @@ pub struct TransitiveOrder {
     /// The effective sell amount for this transient order.
     pub sell: f64,
 }
+
+impl TransitiveOrder {
+    /// Retrieves the exchange price for this order.
+    pub fn price(&self) -> f64 {
+        self.buy / self.sell
+    }
+
+    /// Retrieves the effective price for this order after fees are condidered.
+    ///
+    /// Note that `effective_price > price`.
+    pub fn effective_price(&self) -> f64 {
+        self.price() * FEE_FACTOR
+    }
+}
+
+/// API entry point for computing price estimates and transitive orderbooks for
+/// a give auction.
+pub struct Pricegraph {
+    full_orderbook: Orderbook,
+    reduced_orderbook: Orderbook,
+}
+
+impl Pricegraph {
+    /// Create a new `Pricegraph` instance given an iterator of auction elements
+    /// for the batch.
+    ///
+    /// The auction elements are in the standard exchange format.
+    pub fn new(elements: impl IntoIterator<Item = Element>) -> Self {
+        let orderbook = Orderbook::new(elements);
+        Pricegraph::from_orderbook(orderbook)
+    }
+
+    /// Create a new `Pricegraph` instance from an `Orderbook`.
+    pub fn from_orderbook(mut orderbook: Orderbook) -> Self {
+        let full_orderbook = orderbook.clone();
+        let reduced_orderbook = {
+            orderbook.reduce_overlapping_orders();
+            orderbook
+        };
+
+        Pricegraph {
+            full_orderbook,
+            reduced_orderbook,
+        }
+    }
+
+    /// Gets a copy of the full orderbook for operations that need to contain
+    /// the existing overlapping transitive orders for accuracy.
+    ///
+    /// This method returns a clone of the reduced orderbook because orderbook
+    /// operations are destructive (as they require filling orders).
+    pub fn full_orderbook(&self) -> Orderbook {
+        self.full_orderbook.clone()
+    }
+
+    /// Gets a copy of the reduced orderbook for operations that prefer there to
+    /// be no overlapping transitive orders.
+    ///
+    /// This method returns a clone of the reduced orderbook because orderbook
+    /// operations are destructive (as they require filling orders).
+    pub fn reduced_orderbook(&self) -> Orderbook {
+        self.reduced_orderbook.clone()
+    }
+
+    /// Estimates an exchange rate for the specified token pair and sell volume.
+    /// Returns `None` if the volume cannot be fully filled because there are
+    /// not enough liquidity in the current batch.
+    ///
+    /// Note that this price is in exchange format, that is, it is expressed as
+    /// the ratio between buy and sell amounts, with implicit fees.
```

Comment:
We might want to repeat some documentation from  `fill_market_order` to explain what None and Some(value) mean.

    /// Fill a market order in the current orderbook graph returning the maximum
    /// price the order can have while overlapping with existing orders. Returns
    /// `None` if the order cannot be filled because the token pair is not
    /// connected, or the connection cannot support enough capacity to fully
    /// fill the specified volume.

Or we could mention the documentation of that method.

Reply:
Hmm, I'm not sure how to word this better. I was hoping to keep the "graph" concepts out of the top level API as much as possible and document it from and exchange perspective a little more (so the documentation reads a bit more intuitively). Any suggestions?

Reply:
To simplify the interface, could we return Infinity in case there is not enough liquidity? Or would this just hide a possible error case people need to think about.

Reply:
I think it would return `0` in that case, since its buy / sell, and if there isn't enough liquidity then your buy amount needs to be infinitely small.

Reply:
I'm OK with returning a `0` here if you guys thing it makes more sense.

Reply:
I much prefer Optional over infinity. I slightly prefer Optional over 0 because when talking about exchange rates and prices it doesn't make sense for it to be zero. Not being able to trade is distinct from that which is what Option symbolizes.

Reply:
Any tips on the docs? I was hoping to keep the graph lingo out as much as possible.

Reply:
I *ever so slightly* prefer `Option` as well, as it indicates "there is no exchange rate for what you asked" instead of "here is an invalid exchange rate because it isn't possible".

Reply:
To explain the optional in docs we can say that it will be None if a trade is not possible because no combination of trades is able to trade this amount of the sell token into the buy token.

Reply:
I don't mind *Fill a market order in the current orderbook graph returning the maximum price the order can have while overlapping with existing orders.*. It doesn't even contain graph lingo if you delete the word "graph".

Reply:
`Fill a market order in the current orderbook graph` is not super accurate anymore, because the `orderbook` is always cloned now.

Reply:
I updated it to include your suggestion! I hope it sounds better now :)